src/query/functions/binary/node.go

```diff
@@ -0,0 +1,482 @@
+		GreaterEqType: func(x, y float64) float64 { return toFloat(x >= y) },
+		LesserEqType:  func(x, y float64) float64 { return toFloat(x <= y) },
+	}
+
+	errLeftScalar              = errors.New("expected left scalar but node type incorrect")
+	errRightScalar             = errors.New("expected right scalar but node type incorrect")
+	errNoModifierForComparison = errors.New("comparisons between scalars must use BOOL modifier")
+	errNoMatching              = errors.New("functions must have a vector matching set")
+)
+
+type binaryOp struct {
+	OperatorType string
+	fn           mathFunc
+	info         NodeInformation
+}
+
+// OpType for the operator
+func (o binaryOp) OpType() string {
+	return o.OperatorType
+}
+
+// String representation
+func (o binaryOp) String() string {
+	return fmt.Sprintf("type: %s, lnode: %s, rnode: %s", o.OpType(), o.info.LNode, o.info.RNode)
+}
+
+// Node creates an execution node
+func (o binaryOp) Node(controller *transform.Controller) transform.OpNode {
+	return &binaryNode{
+		controller: controller,
+		cache:      transform.NewBlockCache(),
+		op:         o,
+	}
+}
+
+type binaryNode struct {
+	op         binaryOp
+	controller *transform.Controller
+	cache      *transform.BlockCache
+	mu         sync.Mutex
+}
+
+// NodeInformation describes the types of nodes used
+// for binary operations
+type NodeInformation struct {
+	LNode, RNode         parser.NodeID
+	LIsScalar, RIsScalar bool
+	ReturnBool           bool
+	VectorMatching       *logical.VectorMatching
+}
+
+// NewBinaryOp creates a new binary operation
+func NewBinaryOp(
+	opType string,
+	info NodeInformation,
+) (parser.Params, error) {
+	fn, ok := mathFuncs[opType]
+	if !ok {
+		return binaryOp{}, fmt.Errorf("operator not supported: %s", opType)
+	}
+
+	return binaryOp{
+		OperatorType: opType,
+		// LNode:        lNode,
+		// RNode:        rNode,
+		fn:   fn,
+		info: info,
+	}, nil
+}
+
+// Process processes a block
+func (n *binaryNode) Process(ID parser.NodeID, b block.Block) error {
+	lhs, rhs, err := n.computeOrCache(ID, b)
+	if err != nil {
+		// Clean up any blocks from cache
+		n.cleanup()
+		return err
+	}
+
+	// Both blocks are not ready
+	if lhs == nil || rhs == nil {
+		return nil
+	}
+
+	n.cleanup()
+
+	nextBlock, err := n.process(lhs, rhs)
+	if err != nil {
+		return err
+	}
+
+	defer nextBlock.Close()
+	return n.controller.Process(nextBlock)
+}
+
+// processes two logical blocks, performing a logical operation on them
+func (n *binaryNode) process(lhs, rhs block.Block) (block.Block, error) {
+	info := n.op.info
+
+	lIter, err := lhs.StepIter()
+	if err != nil {
+		return nil, err
+	}
+
+	fn := n.op.fn
+
+	if info.LIsScalar {
+		scalarL, ok := lhs.(*block.ScalarBlock)
+		if !ok {
+			return nil, errLeftScalar
+		}
+
+		lVal := scalarL.Value()
+		// if both lhs and rhs are scalars, can create a new block
+		// by extracting values from lhs and rhs instead of doing
+		// by-value comparisons
+		if info.RIsScalar {
```

Comment:
you can flip the check since the other condition is a single line

Reply:
Good call, holdover from before I pulled out the other functionality into a function